plugins/org.eclipse.n4js/src/org/eclipse/n4js/projectModel/locations/SafeURI.java

```diff
@@ -60,15 +60,14 @@ protected SafeURI(URI wrapped) {
 	protected URI validate(URI given) throws IllegalArgumentException, NullPointerException {
 		Preconditions.checkNotNull(given);
 		List<String> segments = given.segmentsList();
-		if (given.hasTrailingPathSeparator()) {
-			for (int i = 0; i < segments.size() - 1; i++) {
-				Preconditions.checkArgument(segments.get(i).length() > 0, "'%s'", given);
-			}
-		} else {
-			for (String segment : segments) {
-				Preconditions.checkArgument(segment.length() > 0, "'%s'", given);
-			}
+
+		final int segCountMax = given.hasTrailingPathSeparator() ? segments.size() - 1 : segments.size();
+		for (int i = 0; i < segCountMax; i++) {
+			String segment = segments.get(i);
+			Preconditions.checkArgument(segment.length() > 0, "'%s'", given);
+			Preconditions.checkArgument(!segment.contains("\\")); // could happen on Windows platform
```

Comment:
Similar as above. Backslash might appear in a segment on some non-Windows platform.

Reply:
How would a backslash on non-windows (which would not be semantically a slash) translate to a windows path in some URI?

Reply:
In other words, you like to allow backslashes in URIs? (However, not on Windows.)

Reply:
I would have to check how macOS or Linux behave, but in principle a back slash might be a valid character for folder names on some non-Windows platform. In a URI it would have to be escaped, though I'm not even sure that escaping of backslash is strictly required by the standard, but at least it is good practice.

Reply:
I just checked: on macOS backslash is working fine in file/folder names when using the command line (also file name completion with TAB supports it by auto-inserting additional backslashes for escaping).

Reply:
I see, thanks!

Reply:
I have the feeling that this discussion should be about encoded / decoded URIs instead. Would URI.encode / decode handle backslashes correctly?

Reply:
I don't think so, because as far as I understand the code, `Preconditions.checkArgument(!segment.contains("\\"))` is imposing a rule about what characters an individual segment may contain, no matter if they are URI-encoded or not.

BUT: in case `URI#segmentsList()` does not guarantee to return decoded strings (I don't know if it does), then the current code will "overlook" invalid backslashes if they are encoded, which is an unrelated problem (but not critical, I guess).

Reply:
@szarnekow I am not sure how you mean that. Let me explain what I have in my mind.

My intention was to disallow backslashes in URIs on windows (and throw exceptions). The reason for this was that backslashes are forbidden as part of folder or file names on windows. However, as @mor-n4 pointed out, backslashes actually can be part of folder or file names on e.g. MacOS, and hence should be supported on these OSs.

To en-/decode backslashes on Windows would mean that there is a directory named for instance "direc%5Ctory". To create its URI, we use the string "direc\tory" and get after encoding its URI direc%5Ctory"?

Reply:
> My intention was to disallow backslashes in URIs on windows (and throw exceptions). 

Ok, I understand. In that case, I like the last patchset. It looks good to me.